### ISSUE TITLE
Change auth info for the initial ucred to a proper one

### ISSUE THREAD

**Issue by ultimaweapon** [obhq/obliteration#377]
Need to find (or dump?) auth info from the PS4 process that run `exec` on the game binary after forked.

**Comment by ultimaweapon** [comment]
@red-prig do you know the name of (S)ELF file of the process that `exec` the game binary?

**Comment by red-prig** [comment]
@ultimaweapon I suspect that it is somewhere in the depths of vsh, but I have not found the exact location.

**Comment by ultimaweapon** [comment]
Thanks.

**Comment by red-prig** [comment]
@ultimaweapon Here is a log dump from the console, I think in combination with FW research, this can be used to judge the order of initialization
 https://www.psdevwiki.com/ps4/Bootprocess

**Comment by ultimaweapon** [comment]
Thanks.

**Comment by VocalFan** [comment]
Where is this specifically?

**Comment by ultimaweapon** [comment]
It need to be dumped from the kernel somehow. The scene developers should be able to give you some clues.

**Comment by VocalFan** [comment]
okay so, /system/vsh is the location. Will try to snag later :P

**Comment by ultimaweapon** [comment]
> okay so, /system/vsh is the location. Will try to snag later :P

No, it is `/system/sys/SceSysCore.elf`.

Found some evident that it is `SceSysCore.elf`: https://github.com/OpenOrbis/mira-project/issues/154. It is quite small which mean it is likely to be an importance and highly privileged process.

**Comment by VocalFan** [comment]
Eh, I found a SceSysCore.elf in /vsh/ ... Not sure how to snag the /sys/ elf if there is one.

**Comment by VocalFan** [comment]
Uh, email.

**Comment by ultimaweapon** [comment]
That one is `SceShellCore.elf`, not `SceSysCore.elf`.

**Comment by VocalFan** [comment]
...I am the dumb.

**Comment by VocalFan** [comment]
I found a payload that can get authinfo values, will do when home.

**Comment by VocalFan** [comment]
Or uh...
3800000000000007

https://www.psdevwiki.com/ps4/Program_Authority_ID#User_Modules

**Comment by ultimaweapon** [comment]
Strange. Why the value for exe and library does not match with https://github.com/obhq/obliteration/issues/329#issuecomment-1726291728. We still need the value for capabilities and attributes.

**Comment by red-prig** [comment]
It seems that using` sys_get_self_auth_info` you can dump all files in order, I’m also interested in this topic

**Comment by ultimaweapon** [comment]
@red-prig looking good! The problem is it seems like this syscall need to be called from the process we want to dump the auth info.

**Comment by ultimaweapon** [comment]
Wait, seems like the first parameter is the path for the SELF file to get the auth info!

**Comment by ultimaweapon** [comment]
Yeah, I guess the payload you mentioned use this syscall to dump the auth info.

**Comment by VocalFan** [comment]
https://github.com/stooged/Auth-Info-Dumper/blob/master/Auth_Info_Dumper/source/main.c

Can you build a 9.00 payload for me @ultimaweapon ? Latest official build is for 5.05

**Comment by ultimaweapon** [comment]
Seems like it have more works than just compiling. Let me check to see if we can build a simple payload just for `SceSysCore`.

**Comment by VocalFan** [comment]
Alright, because this payload just either kills the GUI, or kernel panics

**Comment by ultimaweapon** [comment]
Required some works because we can't just call `sys_get_self_auth_info` due to the kernel will filter out most of the data we need.

**Comment by VocalFan** [comment]
Alrighty, just keep me updated!

**Comment by ultimaweapon** [comment]
You need to plug the USB before run this. It should create `auth-info.bin` at the root of your USB.

[dump-auth.zip](https://github.com/obhq/obliteration/files/13128278/dump-auth.zip)

**Comment by ultimaweapon** [comment]
Here is what it does:

```c
#include <ps4.h>

struct get_auth_args {
	const char *file;
	unsigned char buf[136];
};

typedef int (*kern_get_self_auth_info_t) (struct thread *, const char *, int, unsigned char *);
typedef int (*copyout_t) (const void *, void *, unsigned);

static int get_auth(struct thread *td, struct get_auth_args *args) {
	void *kernel_base;
	kern_get_self_auth_info_t kern_get_self_auth_info;
	copyout_t copyout;
	unsigned char info[136];
	int error;

	// get kernel functions
	kernel_base = &((uint8_t *)__readmsr(0xC0000082))[-K900_XFAST_SYSCALL];
	kern_get_self_auth_info = (kern_get_self_auth_info_t)(kernel_base + 0x1697B0);
	copyout = (copyout_t)(kernel_base + K900_COPYOUT);

	// get auth
	error = kern_get_self_auth_info(td, args->file, 0, info);

	if (error == 0) {
		error = copyout(info, args->buf, 136);
	}

	return error;
}

int _main(struct thread *td) {
	struct get_auth_args args;
	int fd;

	initKernel();
	initLibc();
	jailbreak();
	initSysUtil();

	// get auth info
	args.file = "/system/sys/SceSysCore.elf";

	if (kexec(get_auth, &args) < 0) {
		printf_notification("Failed to dump auth info.");
		return 0;
	}

	// write to usb drive
	fd = open("/mnt/usb0/auth-info.bin", O_WRONLY | O_CREAT | O_TRUNC, 0777);

	if (fd < 0) {
		printf_notification("Cannot create auth-info.bin on the USB drive.");
		return 0;
	}

	write(fd, args.buf, 136);
	close(fd);

	printf_notification("Auth info dumped!");

	return 0;
}
```

**Comment by VocalFan** [comment]
Will do when home (eta 2 hours ish)

I use a USB to transfer payloads anyways, so the USB requirement is no prob.

**Comment by VocalFan** [comment]
I get "Failed to dump auth info."

**Comment by VocalFan** [comment]
https://github.com/obhq/obliteration/assets/45863583/18e45812-5d8e-4d5d-8a3b-e65efc5eea2b

**Comment by ultimaweapon** [comment]
Try this so see what is the error code.
[dump-auth.zip](https://github.com/obhq/obliteration/files/13160569/dump-auth.zip)

**Comment by VocalFan** [comment]
(2)

**Comment by VocalFan** [comment]
Got hit with an ENOENT? Lol

**Comment by ultimaweapon** [comment]
Strange. Let's try without checking the error.
[dump-auth.zip](https://github.com/obhq/obliteration/files/13160639/dump-auth.zip)

**Comment by VocalFan** [comment]
![image](https://github.com/obhq/obliteration/assets/45863583/465e07cf-75e9-4ba8-8e8e-968dcd145e86)
I don't think this is the intended result.

**Comment by ultimaweapon** [comment]
Does `/system/sys/SceSysCore.elf` exists?

**Comment by VocalFan** [comment]
Lemme check

**Comment by VocalFan** [comment]
![image](https://github.com/obhq/obliteration/assets/45863583/d52f86e6-de3d-481e-bcf5-4babc70dbb25)

**Comment by VocalFan** [comment]
Hey, so, make it try to read SceSysCore.elf off the usb0, I copied the file to my usb's root.

**Comment by VocalFan** [comment]
Ah, so, OpenOrbis believes that the offset isn't correct for get_self_auth_info func in kernel memory.

**Comment by red-prig** [comment]
Although, probably if the application is launched in sandbox mode, the path will be: /[random_word]/sys/SceSysCore.elf 
https://github.com/obhq/obliteration/issues/375

**Comment by red-prig** [comment]
I'm also confused by `unsigned char buf[136];` here you probably need to pass a link to the buffer in the guest code, because it will not return the argument data back to you

**Comment by ultimaweapon** [comment]
> Although, probably if the application is launched in sandbox mode

We already escaped from the sandbox by calling [jailbreak](https://github.com/Scene-Collective/ps4-payload-sdk/blob/master/libPS4/source/payload_utils.c#L112).

> I'm also confused by `unsigned char buf[136];` here you probably need to pass a link to the buffer in the guest code, because it will not return the argument data back to you

Sorry I don't understand what you mean. Could you elaborate this?

**Comment by ultimaweapon** [comment]
I think I found the bug in my code now. The second parameter for `kexec` is not directly passed to the handler. It got wrapped in another struct.

**Comment by red-prig** [comment]
> We already escaped from the sandbox by calling [jailbreak](https://github.com/Scene-Collective/ps4-payload-sdk/blob/master/libPS4/source/payload_utils.c#L112).

Yes, you are jailbreaking, but in what context does the application launch? If it's an imitation of a game, then it's a sandbox.

**Comment by ultimaweapon** [comment]
@VocalFan this should work.

[dump-auth.zip](https://github.com/obhq/obliteration/files/13168042/dump-auth.zip)

**Comment by red-prig** [comment]
> I think I found the bug in my code now. The second parameter for `kexec` is not directly passed to the handler. It got wrapped in another struct.

Yes it is copied.

**Comment by red-prig** [comment]
> > Yes, you are jailbreaking, but in what context does the application launch? If it's an imitation of a game, then it's a sandbox.
> 
> Normally the payload will be executed by the web browser. Not sure what method of Guest Payload that @VocalFan use.

I think `sys_is_in_sandbox` will give an exact answer to the question

**Comment by red-prig** [comment]
> Sorry I don't understand what you mean. Could you elaborate this?

i mean:

```
struct get_auth_args {
	const char *file;
	const char *buf;
};
```

**Comment by red-prig** [comment]
Although if args is copied simply as a link, then for `args->file` you must first do `copyin(args->file, file, 8);` and then `error = kern_get_self_auth_info(td, file, 0, info);`

**Comment by VocalFan** [comment]
> Normally the payload will be executed by the web browser. Not sure what method of Guest Payload that @VocalFan use.

I use AlAzif's Payload Guest, which puts the payload into GoldenHen's payload server... Basically, it has jailbreak/admin access.

**Comment by ultimaweapon** [comment]
@VocalFan see my above comments for the newer version. That one should work.

**Comment by VocalFan** [comment]
I am having the ultimate battle of moving myself from my bed.

**Comment by VocalFan** [comment]
```
07 00 00 00 00 00 00 38 00 00 00 00 00 1C 00 40
00 FF 00 00 00 00 00 80 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 80 00 40 00 40
00 00 00 00 00 00 00 80 00 00 00 00 00 00 00 08
00 40 FF FF 00 00 00 F0 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00
```

**Comment by red-prig** [comment]
Seems true

**Comment by VocalFan** [comment]
> Seems true

Its similar to another SELF's Auth Info

**Comment by VocalFan** [comment]
<pre>
  0000  01 00 00 00 00 10 00 38 00 00 00 00 00 1c 00 40  .......8.......@
  0010  00 ff 00 00 00 00 00 80 00 00 00 00 00 00 00 00  ................
  0020  00 00 00 00 00 00 00 00 00 00 00 c0 00 40 00 40  .............@.@
  0030  00 00 00 00 00 00 00 80 00 00 00 00 00 00 00 f0  ................
  0040  00 40 ff ff 00 00 00 f0 00 00 00 00 00 00 00 00  .@..............
  0050  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
  0060  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
  0070  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
  0080  00 00 00 00 00 00 00 00       
</pre> orbis_swu.self
<pre>
  0000  07 00 00 00 00 00 00 38 00 00 00 00 00 1c 00 40  .......8.......@
  0010  00 ff 00 00 00 00 00 80 00 00 00 00 00 00 00 00  .ÿ.....€........
  0020  00 00 00 00 00 00 00 00 00 00 00 80 00 40 00 40  ...........€.@.@
  0030  00 00 00 00 00 00 00 80 00 00 00 00 00 00 00 08  .......€........
  0040  00 40 ff ff 00 00 00 f0 00 00 00 00 00 00 00 00  .@ÿÿ...ð........
  0050  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
  0060  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
  0070  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
  0080  00 00 00 00 00 00 00 00       
</pre> SceSysCore.self

Note, I added SceSysCore to the Wiki's Auth Info page now, but it's pending approval... Either way, these two SELFs are similar in some ways, different in others.

**Comment by red-prig** [comment]
> ```
> 07 00 00 00 00 00 00 38 00 00 00 00 00 1C 00 40
> 00 FF 00 00 00 00 00 80 00 00 00 00 00 00 00 00
> 00 00 00 00 00 00 00 00 00 00 00 80 00 40 00 40
> 00 00 00 00 00 00 00 80 00 00 00 00 00 00 00 08
> 00 40 FF FF 00 00 00 F0 00 00 00 00 00 00 00 00
> 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
> 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
> 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
> 00 00 00 00 00 00 00 00
> ```
> 
> Or just whole:
> 
> ```
> 07 00 00 00 00 00 00 38 00 00 00 00 00 1C 00 40 00 FF 00 00 00 00 00 80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 80 00 40 00 40 00 00 00 00 00 00 00 80 00 00 00 00 00 00 00 08 00 40 FF FF 00 00 00 F0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
> ```

Judging by: https://www.psdevwiki.com/ps4/Program_Authority_ID#User_Modules 3800000000000007 this is SceSysCore.self so I think everything is correct

**Comment by ultimaweapon** [comment]
@red-prig do you know what is Decid process?

**Comment by red-prig** [comment]
> @red-prig do you know what is Decid process?

This probably has something to do with the debug, I didn't dig into it. Probably the hackers who broke the console can tell you more.

**Comment by ultimaweapon** [comment]
Thanks.

**Comment by VocalFan** [comment]
@ultimaweapon 
![image](https://github.com/obhq/obliteration/assets/45863583/103a9fd4-2683-491d-b5b5-e08480914c8f)
For your working payload ^

**Comment by ultimaweapon** [comment]
It is just a very basic payload with hard coded input. Also it is fixed to 9.0 kernel only. So I don't want to create a repository for a quick and dirty thing.

**Comment by VocalFan** [comment]
> It is just a very basic payload with hard coded input. Also it is fixed to 9.0 kernel only. So I don't want to create a repository for a quick and dirty thing.

But 9.00 is the latest jailbreakable, plus people can just modify the input before compiling.

**Comment by ultimaweapon** [comment]
https://github.com/obhq/auth-dumper
